feat: svelte and form tests

peaklabs-dev opened the pull request:
## Changes

- This is just a demo/ test PR for feedback and some testing.

peaklabs-dev committed:
feat(core): create empty skeleton

peaklabs-dev committed:
feat(lint): custom strict Pint preset

peaklabs-dev committed:
feat(core): New strict `AppServiceProvider.php`

peaklabs-dev committed:
feat(lint): lint full codebase

peaklabs-dev committed:
feat(refactor): Add strict rector configuration

peaklabs-dev committed:
feat(refactor): run rector on the codebase

peaklabs-dev committed:
feat(test): Add peck for typo checking

peaklabs-dev committed:
feat(refactor): improve rector config

- remove withImportNames rule as that is not always what we want
- add a few more Laravel rules

peaklabs-dev committed:
feat(core): New hashing config

- use the much more secure argon2id hashing algorithm (instead of bcrypt)
- adjust the argon2id settings to be more secure

peaklabs-dev committed:
feat(core): improve app config

peaklabs-dev committed:
chore(core): remove unused auth config

peaklabs-dev committed:
feat(test): Add strict ArchTest with custom rules

peaklabs-dev committed:
feat(test): improve ArchTest

- Add a Contract test
- Order the tests better
- Limit usage of Controllers and Requests
- Limit usage of Models
- Forbid the usage of SoftDeletes for Models
- Fix methods names for Queries folder
- Add Services test
- Remove ray as it is covered by the PHP preset

peaklabs-dev committed:
feat(core): New folder structure

peaklabs-dev committed:
feat(core): add more needed folders

peaklabs-dev committed:
feat(core): Add ray to dev for debugging

peaklabs-dev committed:
chore(core): disable ray by default

peaklabs-dev committed:
feat(git): New PR template

peaklabs-dev committed:
feat(git): add funding and assets folder for GitHub

peaklabs-dev committed:
feat(core): New Security document

- New Support Policy table
- Improved reporting guide for vulnerabilities

peaklabs-dev committed:
chore(core): remove css file

- CSS file is not needed for Inertia

peaklabs-dev committed:
chore(core): make sure all pages are ignored in robots.txt

peaklabs-dev committed:
fix(core): wrong import caught by PHPStan

peaklabs-dev committed:
lint(core): lint ray config

peaklabs-dev committed:
chore(core): reorder password rules

peaklabs-dev committed:
feat(core): Add PHPStan with level max

peaklabs-dev committed:
feat(core): Make PHPStan rules even more strict

peaklabs-dev committed:
feat(core): Add testing folders

peaklabs-dev committed:
feat(core): Add PHPStan cache folder

peaklabs-dev committed:
fix(core): ensure force https is always a bool

peaklabs-dev committed:
fix(core): remove `resources` from rector conf

peaklabs-dev committed:
feat(core): new composer.json

- New composer.json with minimal dependencies
- Adds composer scripts for various commands

peaklabs-dev committed:
feat(core): Add back LICENSE with updated year

peaklabs-dev committed:
feat(core): add Bun for Node dependencies

peaklabs-dev committed:
chore(dep): bump node dependencies

peaklabs-dev committed:
fix(dep): lockfile version pinning

peaklabs-dev committed:
chore(dep): bump dep and switch dev command to bun

peaklabs-dev committed:
feat(tests): configure `phpunit.xml` for ArchTest and Pest

peaklabs-dev committed:
chore(core): upgrade phpunit.xml to the latest Laravel skeleton

peaklabs-dev committed:
feat(ui): init tailwind v4 and switch vite to TS

peaklabs-dev committed:
fix(core): invalid config

peaklabs-dev committed:
chore(config): adjust configs to the latest Laravel skeleton

peaklabs-dev committed:
chore(core): use single quotes

peaklabs-dev committed:
feat(core): improve AppServiceProvider

- Added automatic eager loading of relationships
- Configures Laravel Http Facade to prevent stray requests during tests.
- Configures Laravel Sleep Facade to be faked during tests.

peaklabs-dev committed:
feat(core): add base app route

peaklabs-dev committed:
feat(tests): add base test and DB refresh on each test

peaklabs-dev committed:
feat(tests): improve ArchTest

- add more rules for the entire repo
- add Resources rule
- improve existing rules
- adjust formatting

peaklabs-dev committed:
chore(core): .md wording update

peaklabs-dev committed:
fix(types): make types more strict in config files

peaklabs-dev committed:
feat(docker): new strict dockerignore file

peaklabs-dev committed:
chore(docker): adjust dockerignore for more editor files

peaklabs-dev committed:
feat(core): new cleaned up `.env` files

peaklabs-dev committed:
refactor(core): storage framework gitignore file

peaklabs-dev committed:
feat(GitHub): Add new Issue Templates

peaklabs-dev committed:
chore(core): add missing config value & wording update

peaklabs-dev committed:
refactor(workflows): Add optimized stale and lock actions

peaklabs-dev committed:
feat(workflows): Add new test Action which runs all tests

peaklabs-dev committed:
chore(core): rename composer scripts for clarity

peaklabs-dev committed:
chore(workflows): update cache key naming

peaklabs-dev committed:
refactor(core): encrypt session and expire them on close

peaklabs-dev committed:
refactor(core): new v5 changelog

peaklabs-dev committed:
refactor(git): new .gitignore file

andrasbacsai committed:
feat(migrations): add cache and sessions tables with necessary fields

andrasbacsai committed:
feat(inertia): integrate Inertia.js with Laravel, add Vue support and create initial Home page

andrasbacsai committed:
feat(ui): Add shadcn-vue and all related things

andrasbacsai committed:
feat(ui): Add all shadcn components

andrasbacsai committed:
feat(ui): Enhance sidebar component and layout structure with new AppSidebar and Layout.vue files

andrasbacsai committed:
more basic stuffs, like sidebar, teamswitcher, etc.

andrasbacsai committed:
chore(dependencies): add Vue 3.5.13 to package.json and bun.lock

peaklabs-dev committed:
feat(workflows): improve tests action

- add Bun Audit step
- improve naming
- wait on composer validation, bun and composer audit before running test suite

peaklabs-dev committed:
feat: install svelte & upgrade inertia

peaklabs-dev committed:
feat: init svelte

peaklabs-dev committed:
feat: render svelte & vue pages with inertia

peaklabs-dev committed:
feat: share flash messages

peaklabs-dev committed:
feat: svelte from way 1 tanstack

peaklabs-dev committed:
feat: svelte form way 2 tanstack

peaklabs-dev committed:
feat: svelte from way 3 tanstacki - conflicting!

peaklabs-dev committed:
feat: svelte from with new Form component

peaklabs-dev committed:
feat: vue add inertia Form

peaklabs-dev committed:
feat: add vue tanstack from example

peaklabs-dev committed:
feat: add routes

peaklabs-dev commented on resources/js/Pages/Svelte/Way1TanStack.svelte:
```diff
@@ -0,0 +1,218 @@
+                            type="text"
+                            value={field.state.value}
+                            placeholder="John Doe"
+                            oninput={(e: Event) => {
```
Can we maybe remove or replace this by `bind:value={field.state.value}`? Or why do we need this `oninput` handling all the time?

crutchcorn replied:
Without `oninput` we can't track the controlled values via `field.state.value`.

It would be challenging/impossible for us to support all the different bind APIs between Angular, Vue, Svelte, and others

peaklabs-dev replied:
1. That makes sense. Can we somehow extend this on our own to use the bind api?
2. I'm not sure if the Svelte docs are outdated or if this is still supported, but most of the documentation (not the examples) shows this: `onChange={(e) => field.handleChange(e.target.valueAsNumber)}`. This did not work in my testing.

crutchcorn replied:
1. Not easily at the moment - the Form Composition API is the better path forward for Svelte anyway IMO

2. Ahh, yes, `onChange` is wrong for Svelte. We'll update docs

crutchcorn commented on resources/js/Pages/Svelte/Way1TanStack.svelte:
```diff
@@ -0,0 +1,218 @@
+                                Way 2 - Errors from Svelte state submitErrors: {submitErrors.username} <!-- Can we use error mangement form TanStack for this? -->
+
+                                - This can be reset because it is Svelte state.
+                            </p>
+                        {/if}
+                    {/snippet}
+                </form.Field>
+
+                <form.Field name="notifications_enabled">
+                    {#snippet children(field)}
+                        <div class="flex items-center justify-between py-4">
+                            <div class="flex flex-col">
+                                <label
+                                    for={field.name}
+                                    class="text-sm font-medium text-gray-900"
+                                >
+                                    Enable Notifications
+                                </label>
+                                <p class="text-sm text-gray-500">
+                                    Receive notifications about your account
+                                </p>
+                            </div>
+                            <input
+                                id={field.name}
+                                type="checkbox"
+                                checked={field.state.value}
+                                oninput={(e: Event) => {
+                                    const target = e.target as HTMLInputElement;
+                                    field.handleChange(target.checked);
+                                }}
+                                class="h-4 w-4 text-blue-600 border-gray-300 rounded focus:ring-blue-500 focus:ring-2"
+                            />
+                        </div>
+                    {/snippet}
```
We know how verbose this API is on the TanStack Form side.

We have a solution in React/Solid.js (and even Vue, once they fix a bug upstream) but haven't yet ported this API to Svelte:

https://tanstack.com/form/latest/docs/framework/react/guides/form-composition

Dunno how feasible this API is to port to Svelte, but y'all should pop into our Discord and ask - I'll direct the correct maintainers to that question.

peaklabs-dev replied:
Will do.

crutchcorn commented on resources/js/Pages/Svelte/Way1TanStack.svelte:
```diff
@@ -0,0 +1,218 @@
+        flash = {},
+    }: Props = $props(); // Way 1 -> $props() -> Accessing the props of the page via svelte (passed from Inertia).
+
+    // Manual tracking state for this approach. --> Can we use the state management of TanStack and get rid of this?
```
I'm not sure I understand why this is manually being done. `form.state.errorMap` is already a thing in Form - what are we looking to achieve here?

peaklabs-dev replied:
We are tracking this manually because `router.post` from Inertia returns errors if validation does not pass (validation rules on the backend) and we need to catch those errors in order to display them. Is there an easy way after catching these errors to push them into the TanStack errors state?

crutchcorn commented on resources/js/Pages/Svelte/Way1TanStack.svelte:
```diff
@@ -0,0 +1,218 @@
+        },
+        onSubmit: async ({ value }) => {
+            // Way 1 -> Manual form submission via Inertia router.
+            router.post("/test-form", value, {
```
Instead of `onSuccess` try to break apart your APIs to dedicated:

1) Validation step
2) Submission step

And then use `onSubmitAsync`:

https://tanstack.com/form/latest/docs/framework/react/guides/validation#setting-field-level-errors-from-the-forms-validators

peaklabs-dev replied:
This is basically what I want. I want to push errors into TanStack but the problem is that validation happens on submission we don't have separate validation routes.

crutchcorn replied:
In that case, give `form.setErrorMap` a shot instead of `submitErrors`.

peaklabs-dev replied:
I am now using this, and it kind of works.

1. Can I do the same thing if inertia returns successfully or how do we handle that case?
2. Inertia errors come in the format of `errors.{field.name}` which is a problem as I am not sure how to map this smartly to `setErrorMap`
3. Inertia errors have a really good format. I actually prefer there format. They allow accessing errors like `errors.{field.name}`, for example `errors.username` much easier to read then `{field.state.meta.errors[0]}`. Can I do this with TanStack as well? Probably only a type change is required and it would improve error handling a lot.

LeCarbonator replied:
You can use the following structure:
```tsx
const inertiaErrors = {
   username: 'Problem'
}

return {
    fields: inertiaErrors, // cascade the 'Problem' error to the 'username' field
    form: inertiaErrors, // ... but keep a copy on form level
}


form.state.errorMap.onSubmit?.username // 'Problem'
// in field 'username'
field.state.meta.errorMap.onSubmit // 'Problem'
```

peaklabs-dev replied:
Thanks. That is almost working now (see next comment)

crutchcorn reviewed (changes requested):
Some thoughts on Form usage here.

Also, a different route you may want to consider:

JSON Schema Validation

See, one of the main benefits of TanStack Form is our Standard Schema (Zod, Valibot, et al) validation.

It allows us to have stricter TS types, easier validation logic for fields, et al.

To do this with PHP, I'd use:

- https://github.com/swaggest/php-json-schema
- https://github.com/glideapps/zod-from-json-schema

To write JSON schemas on the BE, share them via an isomorphic monorepo file, then use it to generate Zod schemas on the FE.

That's the route I'd go, anyway.

Let me know if you have any additional question - best! 🚀

LeCarbonator commented on resources/js/Pages/Svelte/Way3TanStack.svelte:
```diff
@@ -0,0 +1,199 @@
+    }));
+
+    function handleReset() {
+        form.reset(); // This now has NO effect. So TanStack is not working with the useForm() helper from Inertia?!
```
This may be caused because `$inertiaForm.data` is the most up to date data, and the form will reset to 'the same values' as the default ones.

If the reset should revert any changes made, then the `defaultValues` should not change.

This depends on the behaviour of L19 of course.

peaklabs-dev replied:
This approach is broken because it uses inertia for adding the values `bind:value={$inertiaForm.username}` and this breaks TanStack I think.

LeCarbonator commented on resources/js/Pages/Svelte/Way3TanStack.svelte:
```diff
@@ -0,0 +1,199 @@
+                                type="button"
+                                onclick={handleReset}
+                                disabled={(!state.isDirty &&
+                                    !$inertiaForm.isDirty) ||
```
Does the form helper use non-persistent dirty or persistent dirty?

TanStack Form's `isDirty` does not depend on whether it's the default state or not. That means that once the form has been changed, it can not reach `isDirty` again until it has been reset.
See [the docs](https://tanstack.com/form/latest/docs/framework/react/guides/basic-concepts#understanding-isdirty-in-different-libraries) for more details.

Is that the same strategy that the form helper uses? Or does it compare the initial values with the current ones?

peaklabs-dev replied:
Yes I believe it is the same like TanStack (only after reset `isDirty` is reset).

peaklabs-dev committed:
fix: use setErrorMap

peaklabs-dev commented on resources/js/Pages/Svelte/Way1TanStack.svelte:
```diff
@@ -0,0 +1,221 @@
+                preserveScroll: true,
+                onSuccess: () => {
+                    submitSuccess = true;
+                    form.setErrorMap({}); // is the a metode on TanStack Form to reset errors?
```
Is there a way to reset errors like `clearErrors()` in inertia for example (would be very handy)?

LeCarbonator replied:
The `onSubmit` callback assumes that all validation has been cleared beforehand. Submission validation is cleared if a field changes. 

However, this brings up a good point - perhaps `onSubmit` | `onSubmitAsync` errors should have a configuration for when they should be cleared.

`'clearOnChange', 'clearOnSubmit'`, any other ones that would fit?

peaklabs-dev replied:
Yes having `clearOnChange`, `clearOnSubmit` and maybe more like `clearOnBlur` would be highly helpful for adjusting this behavior

- `clearOnChange` -> Clear on change of the field (default now)
- `clearOnSubmit` ->  Clear errors before submitting and then submit (because submit can error again if using inertia for example).

peaklabs-dev committed:
feat: TanStack from with inertia state

peaklabs-dev committed:
fix: tanstack error states

peaklabs-dev commented on resources/js/Pages/Svelte/Way1TanStack.svelte:
```diff
@@ -0,0 +1,227 @@
+                    form.setErrorMap({
+                        onSubmit: {
+                            fields: errors,
+                            form: errors,
```
1. This form level option does not seem to work, as form-level errors are not set here. Because when I do `form.state.errors.length > 0` this is 0 even with errors. How do I properly set the errors no only at field level but also at form level?

peaklabs-dev replied:
2. There is also a second problem with this. I want to use a `zod` schema for `onChange` or `onBlur` validation but `onSubmit` I want to clear all errors and populate the form with the server-side errors. Currently, when there are errors (from `zod`) and I hit submit (I think nothing happens) it does not seem like the server-side validation errors come through.
-> To fix this I would like to clear the errors `onChange` and then revalidate using `zod` `onChange` and `onSubmit` I need to clear all errors before submitting to make sure after submit now my server side errors are used (see this for more detail https://github.com/coollabsio/coolify/pull/6389#discussion_r2285563804)

peaklabs-dev commented on resources/js/Pages/Svelte/Way1TanStack.svelte:
```diff
@@ -0,0 +1,227 @@
+            router.post("/test-form", value, {
+                preserveScroll: true,
+                onSuccess: () => {
+                    submitSuccess = true;
```
Can I let TanStack know here that it was a `success` instead of manually tracking it?

peaklabs-dev committed:
remove not needed comments

peaklabs-dev committed:
feat(core): add temporary Contributing guide

peaklabs-dev committed:
feat(core): improve local dev until docker is ready

peaklabs-dev committed:
Merge branch 'v5.x' into v5.x-form-test

peaklabs-dev committed:
feat: add route for state test

peaklabs-dev commented on resources/js/Pages/Svelte/Way1TanStack.svelte:
```diff
@@ -0,0 +1,227 @@
+                                class="h-4 w-4 text-blue-600 border-gray-300 rounded focus:ring-blue-500 focus:ring-2"
+                            />
+                        </div>
+                        {#if field.state.meta.errors.length > 0}
```
What is the difference between this way or `field.state.meta.isValid` -> Is `isValid` just set to false if there are errors on the field?

Also the docs has a mistake which cost me quite some time to notice: `{#if field.state.meta.errors}` is used as a condition which does not work even when there are no errors this is because because empty arrays are `true`.

LeCarbonator replied:
`field.state.meta.isValid === field.state.meta.errors.length === 0`

Good catch with the docs! Could you open an issue for it?

peaklabs-dev commented on resources/js/Pages/Svelte/Way1TanStack.svelte:
```diff
@@ -0,0 +1,227 @@
+                    // Inertia: errors.username = "error message"
+                    form.setErrorMap({
+                        onSubmit: {
+                            fields: errors,
```
This option also does not work completely. If multiple errors are returned only 1 is kept and shown in the UI https://laravel.com/docs/12.x/validation#validation-error-response-format. And if I submit again the state is stuck and only if I reload the page, fix the first error (from my mind, what I changed before) the second error appears (after the first one is solved).

peaklabs-dev commented on resources/js/Pages/Svelte/Way1TanStack.svelte:
```diff
@@ -0,0 +1,227 @@
+                        {#if field.state.meta.errors.length > 0}
+                            <!-- Using TanStack Form's built-in error management -->
+                            <p class="mt-2 text-sm text-red-600">
+                                TanStack Form Errors: {field.state.meta.errors}
```
I tried `{field.state.meta.errors.join(', ')`} and this and both only show 1 error message.

peaklabs-dev commented on resources/js/Pages/Svelte/Way1TanStackWithState.svelte:
```diff
@@ -0,0 +1,219 @@
+    const form = createForm(() => ({
+        defaultValues: restoredState,
+        onSubmit: async ({ value }) => {
+            router.post("/test-form", value, { // Now we have a type Error here??
```
There is now a type error here, probably because of the default values.

peaklabs-dev commented on resources/js/Pages/Svelte/Way1TanStackWithState.svelte:
```diff
@@ -0,0 +1,219 @@
+    let submitSuccess = $state(false);
+
+    const form = createForm(() => ({
+        defaultValues: restoredState,
```
I want to restore the full state but I need a better way to restore the full state then hijacking the `defaultValues`. Can I not restore the full state directly into `formApi.state.values` on line 15 or so?

peaklabs-dev committed:
feat: add state debugging

peaklabs-dev committed:
chore(deps): bump packages

peaklabs-dev committed:
feat: add zod schema

peaklabs-dev committed:
chore: improve routes

- use inertia() helper
- add additional validation rule.
- simulate a delay of the server response.

peaklabs-dev committed:
feat: add type to inertia router.restore

peaklabs-dev committed:
feat: improve TanStack From

- use onDynamic validation. Which validates onChange after the first submit.
- use zod schema
- await inertia router call
- fix state issue with svelte

peaklabs-dev committed:
cleanup

peaklabs-dev committed:
feat: add useForm example

peaklabs-dev committed:
cleanup
